processing/src/main/java/org/apache/druid/segment/RowWalker.java

```diff
@@ -60,13 +61,16 @@ public T currentRow()
 
   public void advance()
   {
-    if (rowIterator.hasNext()) {
+    if (rowIterator == null) {
+      throw new IllegalStateException("cannot call advance when isDone == true");
```

Comment:
Note to reviewers: adding this check, without the patch to HashJoinEngine, causes a few of the tests to fail and illustrates the problem where `advance` is called on a cursor that has no more rows left.